xskillscore/tests/test_deterministic.py

```diff
@@ -5,229 +5,341 @@
+
+    actual = pearson_r_p_value(a, b, dim, weights=_weights)
     assert actual.chunks is None
 
-    dim, _ = _preprocess(dim)
+    dim, _ = _preprocess_dims(dim)
     if len(dim) > 1:
-        new_dim = '_'.join(dim)
+        new_dim = "_".join(dim)
         _a = a.stack(**{new_dim: dim})
         _b = b.stack(**{new_dim: dim})
+        if weight:
+            _weights = _weights.stack(**{new_dim: dim})
     else:
         new_dim = dim[0]
         _a = a
         _b = b
+    _weights = _preprocess_weights(_a, dim, new_dim, _weights)
 
     axis = _a.dims.index(new_dim)
-    res = _pearson_r_p_value(_a.values, _b.values, axis)
+    res = _pearson_r_p_value(_a.values, _b.values, _weights.values, axis)
     expected = actual.copy()
     expected.values = res
     assert_allclose(actual, expected)
 
 
-@pytest.mark.parametrize('dim', AXES)
-def test_pearson_r_p_value_xr_dask(a_dask, b_dask, dim):
-    actual = pearson_r_p_value(a_dask, b_dask, dim)
+@pytest.mark.parametrize("dim", AXES)
+@pytest.mark.parametrize("weight", [True, False])
+def test_pearson_r_p_value_xr_dask(a_dask, b_dask, dim, weight, weights_dask):
+    # Generates subsetted weights to pass in as arg to main function and for the numpy testing.
+    _weights = adjust_weights(dim, weight, weights_dask)
+
+    actual = pearson_r_p_value(a_dask, b_dask, dim, weights=_weights)
     assert actual.chunks is not None
 
-    dim, _ = _preprocess(dim)
+    dim, _ = _preprocess_dims(dim)
     if len(dim) > 1:
-        new_dim = '_'.join(dim)
+        new_dim = "_".join(dim)
         _a_dask = a_dask.stack(**{new_dim: dim})
         _b_dask = b_dask.stack(**{new_dim: dim})
+        if weight:
+            _weights = _weights.stack(**{new_dim: dim})
     else:
         new_dim = dim[0]
         _a_dask = a_dask
         _b_dask = b_dask
+    _weights = _preprocess_weights(_a_dask, dim, new_dim, _weights)
 
     axis = _a_dask.dims.index(new_dim)
-    res = _pearson_r_p_value(_a_dask.values, _b_dask.values, axis)
+    res = _pearson_r_p_value(_a_dask.values, _b_dask.values, _weights.values, axis)
     expected = actual.copy()
     expected.values = res
     assert_allclose(actual, expected)
 
 
-@pytest.mark.parametrize('dim', AXES)
-def test_rmse_r_xr(a, b, dim):
-    actual = rmse(a, b, dim)
+@pytest.mark.parametrize("dim", AXES)
+@pytest.mark.parametrize("weight", [True, False])
+def test_rmse_xr(a, b, dim, weight, weights):
+    # Generates subsetted weights to pass in as arg to main function and for the numpy testing.
+    weights = adjust_weights(dim, weight, weights)
+
+    actual = rmse(a, b, dim, weights=weights)
     assert actual.chunks is None
 
-    dim, axis = _preprocess(dim)
-    _a = a.values
-    _b = b.values
+    dim, axis = _preprocess_dims(dim)
+    _a = a
+    _b = b
+    _weights = _preprocess_weights(_a, dim, dim, weights)
     axis = tuple(a.dims.index(d) for d in dim)
-    res = _rmse(_a, _b, axis)
+    res = _rmse(_a.values, _b.values, _weights, axis)
```

Comment:
Does ``_weights`` need to be ``_weights.values``?

Reply:
I just updated it. It works either way. Originally I was passing a single `None` into here so I couldn't do `.values` but now it passes an array like `a` or `b` with `None`s so this works now.